File: app/src/main/java/com/nruge/iceinfo/TrainRepository.kt
package com.nruge.iceinfo

import com.nruge.iceinfo.model.*
import com.nruge.iceinfo.util.calculateDelayMinutes
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.okhttp.*
import io.ktor.client.plugins.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.logging.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import android.util.Log
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

object TrainRepository {

    private const val API_PATH_STATUS = "/api1/rs/status"
    private const val API_PATH_TRIP   = "/api1/rs/tripInfo/trip"
    private const val API_PATH_POIS   = "/api1/rs/pois/map"
    private const val API_PATH_CONN   = "/api1/rs/tripInfo/connection"

    // Try HTTPS first; many older ICE portals also serve plain HTTP
    private val hosts = listOf("https://iceportal.de", "http://iceportal.de")

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    private val client = HttpClient(OkHttp) {
        install(ContentNegotiation) {
            json(json, contentType = io.ktor.http.ContentType.Any)
        }
        install(Logging) {
            level = LogLevel.INFO
        }
        install(HttpTimeout) {
            requestTimeoutMillis = 5000
            connectTimeoutMillis = 3000
            socketTimeoutMillis  = 5000
        }
        defaultRequest {
            header("Accept", "application/json")
        }
    }

    /** GET with automatic HTTP fallback if HTTPS fails. */
    private suspend inline fun <reified T> getWithFallback(path: String): T {
        var lastException: Exception? = null
        for (host in hosts) {
            try {
                return client.get("$host$path").body()
            } catch (e: Exception) {
                Log.w("ICERepo", "GET $host$path failed: ${e.message}")
                lastException = e
            }
        }
        throw lastException ?: IllegalStateException("No hosts configured for $path")
    }

    /** Raw text GET with automatic HTTP fallback. */
    private suspend fun getRawWithFallback(path: String): String {
        var lastException: Exception? = null
        for (host in hosts) {
            try {
                return client.get("$host$path").bodyAsText()
            } catch (e: Exception) {
                Log.w("ICERepo", "GET $host$path failed: ${e.message}")
                lastException = e
            }
        }
        throw lastException ?: IllegalStateException("No hosts configured for $path")
    }

    private val timeFormatter: DateTimeFormatter =
        DateTimeFormatter.ofPattern("HH:mm").withZone(ZoneId.systemDefault())

    suspend fun fetchPois(lat: Double, lon: Double, radiusDeg: Double = 0.5): List<PoiItem> = withContext(Dispatchers.IO) {
        if (lat == 0.0 && lon == 0.0) return@withContext emptyList()
        val path = "$API_PATH_POIS/${lat - radiusDeg}/${lon - radiusDeg}/${lat + radiusDeg}/${lon + radiusDeg}"
        try {
            // The ICE portal returns either {"pois":[…]} or a bare array [… ] depending on firmware.
            val raw = getRawWithFallback(path)
            val element = json.parseToJsonElement(raw)
            when (element) {
                is JsonArray  -> json.decodeFromJsonElement<List<PoiItem>>(element)
                is JsonObject -> {
                    val inner = element["pois"] ?: return@withContext emptyList()
                    json.decodeFromJsonElement<List<PoiItem>>(inner)
                }
                else -> emptyList()
            }.sortedBy { it.distance }
        } catch (e: Exception) {
            Log.e("ICERepo", "POI Fehler ($path): ${e.message}", e)
            emptyList()
        }
    }

    suspend fun fetchTrainStatus(): TrainStatus = withContext(Dispatchers.IO) {
        try {
            val status: StatusResponse = getWithFallback(API_PATH_STATUS)
            val tripResponse: TripResponse = getWithFallback(API_PATH_TRIP)
            val trip = tripResponse.trip ?: return@withContext fallback()
            mapToTrainStatus(status, trip)
        } catch (e: Exception) {
            Log.e("ICERepo", "Status Fehler: ${e.message}", e)
            fallback()
        }
    }

    private fun mapToTrainStatus(status: StatusResponse, trip: TripInfo): TrainStatus {
        val stops = trip.stops
        val lastStop = stops.lastOrNull()
        val destination = lastStop?.station?.name ?: "Unbekannt"
        val destTimetable = lastStop?.timetable
        val destScheduledMs = destTimetable?.scheduledArrivalTime ?: 0L
        val destActualMs = destTimetable?.actualArrivalTime ?: 0L
        val destinationEta = formatTime(destScheduledMs)
        val destinationTrack = lastStop?.track?.actual ?: ""
        val destinationDelay = calculateDelayMinutes(destActualMs, destScheduledMs)
        val totalDistance = lastStop?.info?.distanceFromStart ?: 0
        val currentDistance = trip.actualPosition
        val distanceToDestination = totalDistance - currentDistance
        var nextStopName = "Unbekannt"
        var eta = "--:--"
        var delayMinutes = 0
        var track = ""
        var delayReason = ""
        var distanceToNext = 0
        var distanceLastToNext = 0
        val stopList = mutableListOf<TrainStop>()
        var nextFound = false
        var nextStopEva = ""

        stops.forEachIndexed { i, stop ->
            val info = stop.info ?: return@forEachIndexed
            val passed = info.passed
            val timetable = stop.timetable

            val scheduledMs = timetable?.scheduledArrivalTime ?: 0L
            val actualMs = timetable?.actualArrivalTime ?: 0L
            val stopDelay = calculateDelayMinutes(actualMs, scheduledMs)

            val depScheduledMs = timetable?.scheduledDepartureTime ?: 0L
            val depActualMs = timetable?.actualDepartureTime ?: 0L
            val depDelay = calculateDelayMinutes(depActualMs, depScheduledMs)

            val stopTrack = stop.track?.actual ?: ""
            val stopName = stop.station?.name ?: "?"

            val isNext = !passed && !nextFound
            if (isNext) {
                nextStopEva = stop.station?.evaNr ?: ""
                nextFound = true
                nextStopName = stopName
                eta = formatTime(scheduledMs)
                delayMinutes = stopDelay
                track = stopTrack
                distanceToNext = info.distance

                val distanceFromStart = info.distanceFromStart
                distanceLastToNext = distanceFromStart - (if (i > 0) {
                    stops[i - 1].info?.distanceFromStart ?: 0
                } else 0)

                delayReason = stop.delayReasons?.firstOrNull()?.text ?: ""
            }

            stopList.add(TrainStop(
                name = stopName,
                evaNr = stop.station?.evaNr ?: "",
                scheduledArrival = formatTime(scheduledMs),
                actualArrival = formatTime(actualMs),
                delayMinutes = stopDelay,
                track = stopTrack,
                passed = passed,
                isNext = isNext,
                distanceFromStart = info.distanceFromStart,
                scheduledArrivalMs = scheduledMs,
                isAdditional = info.status == 2,
                scheduledDeparture = formatTime(depScheduledMs),
                actualDeparture = formatTime(depActualMs),
                departureDelayMinutes = depDelay,
                isCancelled = stop.cancelled || info.status == 3
            ))
        }

        return TrainStatus(
            trainType = trip.trainType,
            trainNumber = trip.vzn,
            speed = status.speed.toInt(),
            nextStop = nextStopName,
            destination = destination,
            nextStopEva = nextStopEva,
            eta = eta,
            delayMinutes = delayMinutes,
            track = track,
            delayReason = delayReason,
            distanceToNext = distanceToNext,
            distanceLastToNext = distanceLastToNext,
            stops = stopList,
            wagonClass = status.wagonClass,
            connectivity = status.connectivity?.currentState ?: "",
            nextConnectivity = status.connectivity?.nextState,
            connectivityRemainingSeconds = status.connectivity?.remainingTimeSeconds,
            tzn = status.tzn,
            latitude = status.latitude,
            longitude = status.longitude,
            distanceToDestination = distanceToDestination,
            actualPosition = trip.actualPosition,
            destinationEta = destinationEta,
            destinationTrack = destinationTrack,
            destinationDelay = destinationDelay,
            isConnected = true
        )
    }

    suspend fun fetchConnections(evaNr: String, ourArrivalMs: Long = 0L): List<ConnectingTrain> = withContext(Dispatchers.IO) {
        try {
            if (evaNr.isEmpty()) return@withContext emptyList()
            val response: ConnectionResponse = getWithFallback("$API_PATH_CONN/$evaNr")
            val now = System.currentTimeMillis()
            response.connections?.map { c ->
                val scheduledMs = c.timetable?.scheduledDepartureTime ?: 0L
                val actualMs = c.timetable?.actualDepartureTime ?: 0L
                val delayMin = calculateDelayMinutes(actualMs, scheduledMs)
                val effectiveDepartureMs = if (actualMs > 0) actualMs else scheduledMs
                val reachable = when {
                    effectiveDepartureMs <= 0L -> true
                    ourArrivalMs > 0L -> effectiveDepartureMs > ourArrivalMs
                    else -> effectiveDepartureMs > now
                }
                val transferMinutes = if (ourArrivalMs > 0L && effectiveDepartureMs > ourArrivalMs) {
                    ((effectiveDepartureMs - ourArrivalMs) / 60_000L).toInt()
                } else null
                ConnectingTrain(
                    trainType = c.trainType,
                    trainNumber = c.vzn,
                    destination = c.finalStation,
                    departure = formatTime(scheduledMs),
                    track = c.track?.actual ?: "",
                    delayMinutes = delayMin,
                    reachable = reachable,
                    transferMinutes = transferMinutes
                )
            } ?: emptyList()
        } catch (e: Exception) {
            Log.e("ICERepo", "Connection Fehler: ${e.message}")
            emptyList()
        }
    }

    data class DebugData(
        val tripRaw: String,
        val tripError: String?,
        val connectionRaw: String,
        val connectionError: String?,
        val evaNr: String
    )

    suspend fun fetchDebugData(): DebugData = withContext(Dispatchers.IO) {
        var tripRaw = ""
        var tripError: String? = null
        var evaNr = ""

        try {
            tripRaw = getRawWithFallback(API_PATH_TRIP)
            try {
                val tripResponse = json.decodeFromString<TripResponse>(tripRaw)
                evaNr = tripResponse.trip?.stops
                    ?.firstOrNull { it.info?.passed == false }
                    ?.station?.evaNr ?: ""
            } catch (_: Exception) {}
        } catch (e: Exception) {
            tripError = e.message ?: "Unbekannter Fehler"
        }

        var connectionRaw = ""
        var connectionError: String? = null
        if (evaNr.isNotEmpty()) {
            try {
                connectionRaw = getRawWithFallback("$API_PATH_CONN/$evaNr")
            } catch (e: Exception) {
                connectionError = e.message ?: "Unbekannter Fehler"
            }
        } else {
            connectionError = "EVA-Nummer nicht verfügbar"
        }

        DebugData(tripRaw, tripError, connectionRaw, connectionError, evaNr)
    }

    private fun formatTime(ms: Long): String {
        if (ms <= 0L) return ""
        return timeFormatter.format(Instant.ofEpochMilli(ms))
    }

    private fun fallback() = sampleTrainStatus.copy(isConnected = false)
}

File: ios/ICEInfo/Network/TrainRepository.swift
import Foundation

actor TrainRepository {
    static let shared = TrainRepository()

    private let hosts = ["https://iceportal.de", "http://iceportal.de"]
    private let session: URLSession
    private let decoder: JSONDecoder
    private let dateFormatter: DateFormatter

    private let pathStatus = "/api1/rs/status"
    private let pathTrip = "/api1/rs/tripInfo/trip"
    private let pathPois = "/api1/rs/pois/map"
    private let pathConn = "/api1/rs/tripInfo/connection"

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 5
        config.timeoutIntervalForResource = 5
        config.requestCachePolicy = .reloadIgnoringLocalCacheData
        session = URLSession(configuration: config)

        decoder = JSONDecoder()

        dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "HH:mm"
        dateFormatter.timeZone = TimeZone.current
    }

    private func getWithFallback<T: Decodable & Sendable>(path: String) async throws -> T {
        var lastError: Error? = nil
        for host in hosts {
            guard let url = URL(string: "\(host)\(path)") else { continue }
            do {
                var request = URLRequest(url: url)
                request.setValue("application/json", forHTTPHeaderField: "Accept")
                let (data, response) = try await session.data(for: request)
                guard let httpResponse = response as? HTTPURLResponse else {
                    throw APIError.noData
                }
                guard (200...299).contains(httpResponse.statusCode) else {
                    throw APIError.httpError(httpResponse.statusCode)
                }
                return try decoder.decode(T.self, from: data)
            } catch {
                lastError = error
                continue
            }
        }
        throw lastError ?? APIError.allHostsFailed
    }

    private func getRawWithFallback(path: String) async throws -> String {
        var lastError: Error? = nil
        for host in hosts {
            guard let url = URL(string: "\(host)\(path)") else { continue }
            do {
                var request = URLRequest(url: url)
                request.setValue("application/json", forHTTPHeaderField: "Accept")
                let (data, response) = try await session.data(for: request)
                guard let httpResponse = response as? HTTPURLResponse else {
                    throw APIError.noData
                }
                guard (200...299).contains(httpResponse.statusCode) else {
                    throw APIError.httpError(httpResponse.statusCode)
                }
                return String(data: data, encoding: .utf8) ?? ""
            } catch {
                lastError = error
                continue
            }
        }
        throw lastError ?? APIError.allHostsFailed
    }

    func fetchPois(lat: Double, lon: Double, radiusDeg: Double = 0.5) async -> [PoiItem] {
        guard !(lat == 0 && lon == 0) else { return [] }
        let path = "\(pathPois)/\(lat - radiusDeg)/\(lon - radiusDeg)/\(lat + radiusDeg)/\(lon + radiusDeg)"
        do {
            let raw = try await getRawWithFallback(path: path)
            guard let data = raw.data(using: .utf8) else { return [] }

            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
               let poisArray = json["pois"] as? [[String: Any]] {
                let poisData = try JSONSerialization.data(withJSONObject: poisArray)
                var items = try decoder.decode([PoiItem].self, from: poisData)
                items.sort { $0.distance < $1.distance }
                return items
            } else if let poisArray = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] {
                let poisData = try JSONSerialization.data(withJSONObject: poisArray)
                var items = try decoder.decode([PoiItem].self, from: poisData)
                items.sort { $0.distance < $1.distance }
                return items
            }
            return []
        } catch {
            print("POI error: \(error)")
            return []
        }
    }

    func fetchTrainStatus() async -> TrainStatus {
        do {
            async let statusResult: StatusResponse = getWithFallback(path: pathStatus)
            async let tripResult: TripResponse = getWithFallback(path: pathTrip)
            let (status, tripResponse) = try await (statusResult, tripResult)
            guard let trip = tripResponse.trip else { return disconnectedStatus() }
            return mapToTrainStatus(status: status, trip: trip)
        } catch {
            print("Status error: \(error)")
            return disconnectedStatus()
        }
    }

    func fetchDebugData() async -> (tripRaw: String, tripError: String?, connectionRaw: String, connectionError: String?, evaNr: String) {
        var tripRaw = ""
        var tripError: String? = nil
        var evaNr = ""
        do {
            tripRaw = try await getRawWithFallback(path: pathTrip)
            if let data = tripRaw.data(using: .utf8),
               let tripResponse = try? decoder.decode(TripResponse.self, from: data) {
                evaNr = tripResponse.trip?.stops.first(where: { $0.info?.passed == false })?.station?.evaNr ?? ""
            }
        } catch {
            tripError = error.localizedDescription
        }

        var connectionRaw = ""
        var connectionError: String? = nil
        if !evaNr.isEmpty {
            do {
                connectionRaw = try await getRawWithFallback(path: "\(pathConn)/\(evaNr)")
            } catch {
                connectionError = error.localizedDescription
            }
        } else {
            connectionError = "EVA number not available"
        }

        return (tripRaw, tripError, connectionRaw, connectionError, evaNr)
    }

    func fetchConnections(evaNr: String, ourArrivalMs: Int64 = 0) async -> [ConnectingTrain] {
        guard !evaNr.isEmpty else { return [] }
        do {
            let response: ConnectionResponse = try await getWithFallback(path: "\(pathConn)/\(evaNr)")
            let now = Date().timeIntervalSince1970 * 1000
            return response.connections?.map { conn in
                let scheduledMs = conn.timetable?.scheduledDepartureTime ?? 0
                let actualMs = conn.timetable?.actualDepartureTime ?? 0
                let delayMin = calculateDelayMinutes(actual: actualMs, scheduled: scheduledMs)
                let effectiveDepartureMs = actualMs > 0 ? actualMs : scheduledMs
                let reachable: Bool
                if effectiveDepartureMs <= 0 {
                    reachable = true
                } else if ourArrivalMs > 0 {
                    reachable = effectiveDepartureMs > ourArrivalMs
                } else {
                    reachable = effectiveDepartureMs > Int64(now)
                }
                let transferMinutes: Int?
                if ourArrivalMs > 0 && effectiveDepartureMs > ourArrivalMs {
                    transferMinutes = Int((effectiveDepartureMs - ourArrivalMs) / 60_000)
                } else {
                    transferMinutes = nil
                }
                return ConnectingTrain(
                    trainType: conn.trainType,
                    trainNumber: conn.vzn,
                    destination: conn.finalStation,
                    departure: formatTime(ms: scheduledMs),
                    track: conn.track?.actual ?? "",
                    delayMinutes: delayMin,
                    reachable: reachable,
                    transferMinutes: transferMinutes
                )
            } ?? []
        } catch {
            print("Connection error: \(error)")
            return []
        }
    }

    private func mapToTrainStatus(status: StatusResponse, trip: TripInfo) -> TrainStatus {
        let stops = trip.stops
        let lastStop = stops.last
        let destination = lastStop?.station?.name ?? "Unbekannt"
        let destTimetable = lastStop?.timetable
        let destScheduledMs = destTimetable?.scheduledArrivalTime ?? 0
        let destActualMs = destTimetable?.actualArrivalTime ?? 0
        let destinationEta = formatTime(ms: destScheduledMs)
        let destinationTrack = lastStop?.track?.actual ?? ""
        let destinationDelay = calculateDelayMinutes(actual: destActualMs, scheduled: destScheduledMs)
        let totalDistance = lastStop?.info?.distanceFromStart ?? 0
        let currentDistance = trip.actualPosition
        let distanceToDestination = totalDistance - currentDistance

        var nextStopName = "Unbekannt"
        var eta = "--:--"
        var delayMinutes = 0
        var track = ""
        var delayReason = ""
        var distanceToNext = 0
        var distanceLastToNext = 0
        var stopList: [TrainStop] = []
        var nextFound = false
        var nextStopEva = ""

        for (i, stop) in stops.enumerated() {
            guard let info = stop.info else { continue }
            let passed = info.passed
            let timetable = stop.timetable

            let scheduledMs = timetable?.scheduledArrivalTime ?? 0
            let actualMs = timetable?.actualArrivalTime ?? 0
            let stopDelay = calculateDelayMinutes(actual: actualMs, scheduled: scheduledMs)

            let depScheduledMs = timetable?.scheduledDepartureTime ?? 0
            let depActualMs = timetable?.actualDepartureTime ?? 0
            let depDelay = calculateDelayMinutes(actual: depActualMs, scheduled: depScheduledMs)

            let stopTrack = stop.track?.actual ?? ""
            let stopName = stop.station?.name ?? "?"

            let isNext = !passed && !nextFound
            if isNext {
                nextStopEva = stop.station?.evaNr ?? ""
                nextFound = true
                nextStopName = stopName
                eta = formatTime(ms: scheduledMs)
                delayMinutes = stopDelay
                track = stopTrack
                distanceToNext = info.distance

                let distanceFromStart = info.distanceFromStart
                distanceLastToNext = distanceFromStart - (i > 0 ? (stops[i - 1].info?.distanceFromStart ?? 0) : 0)

                delayReason = stop.delayReasons?.first?.text ?? ""
            }

            stopList.append(TrainStop(
                name: stopName,
                evaNr: stop.station?.evaNr ?? "",
                scheduledArrival: formatTime(ms: scheduledMs),
                actualArrival: formatTime(ms: actualMs),
                delayMinutes: stopDelay,
                track: stopTrack,
                passed: passed,
                isNext: isNext,
                distanceFromStart: info.distanceFromStart,
                scheduledArrivalMs: scheduledMs,
                scheduledDepartureMs: depScheduledMs,
                isAdditional: info.status == 2,
                scheduledDeparture: formatTime(ms: depScheduledMs),
                actualDeparture: formatTime(ms: depActualMs),
                departureDelayMinutes: depDelay,
                isCancelled: stop.cancelled || info.status == 3
            ))
        }

        return TrainStatus(
            trainType: trip.trainType,
            trainNumber: trip.vzn,
            speed: Int(status.speed),
            nextStop: nextStopName,
            destination: destination,
            eta: eta,
            delayMinutes: delayMinutes,
            track: track,
            delayReason: delayReason,
            distanceToNext: distanceToNext,
            distanceLastToNext: distanceLastToNext,
            nextStopEva: nextStopEva,
            stops: stopList,
            wagonClass: status.wagonClass,
            connectivity: status.connectivity?.currentState ?? "",
            nextConnectivity: status.connectivity?.nextState,
            connectivityRemainingSeconds: status.connectivity?.remainingTimeSeconds,
            tzn: status.tzn,
            series: status.series,
            latitude: status.latitude,
            longitude: status.longitude,
            distanceToDestination: distanceToDestination,
            actualPosition: trip.actualPosition,
            destinationEta: destinationEta,
            destinationTrack: destinationTrack,
            destinationDelay: destinationDelay,
            isConnected: true
        )
    }

    private func formatTime(ms: Int64) -> String {
        guard ms > 0 else { return "" }
        let date = Date(timeIntervalSince1970: TimeInterval(ms) / 1000)
        return dateFormatter.string(from: date)
    }

    private func disconnectedStatus() -> TrainStatus {
        var status = sampleTrainStatus
        status.isConnected = false
        return status
    }
}

func calculateDelayMinutes(actual: Int64, scheduled: Int64) -> Int {
    guard actual > 0 && scheduled > 0 else { return 0 }
    return Int((actual - scheduled) / 60_000)
}
